source/mir/deser/msgpack.d

```diff
@@ -7,34 +7,864 @@ Macros:
 IONREF = $(REF_ALTTEXT $(TT $2), $2, mir, ion, $1)$(NBSP)
 +/
 module mir.deser.msgpack;
+import mir.algebraic : Nullable;
+import mir.ser.msgpack : MessagePackFmt;
+import mir.ion.exception;
+import mir.lob : Blob;
 
-version(Have_msgpack_d)
+struct MsgpackExtension
 {
-    /++
-    +/
-    T deserializeMsgpack(T)(scope const(ubyte)[] data)
+    Blob data;
+    ubyte type;
+}
+
+struct MsgpackList
+{
+    const(ubyte)[] data;
+    size_t length;
+}
+
+struct MsgpackStruct
+{
+    const(ubyte)[] data;
+    size_t length;
+}
+
+struct MsgpackDescribedValue
+{
+    const(ubyte)[] data;
+    MessagePackFmt type;
+}
+
+private static T unpackMsgPackVal(T)(scope ref const(ubyte)[] data)
+{
+    import std.traits : Unsigned;
+    alias UT = Unsigned!T;
+
+    if (data.length < UT.sizeof)
+    {
+        version (D_Exceptions)
+            throw IonErrorCode.unexpectedEndOfData.ionException;
+        else
+            assert(0, IonErrorCode.unexpectedEndOfData.ionErrorMsg);
+    }
+
+    UT ret = (cast(UT[1])cast(ubyte[UT.sizeof])data[0 .. UT.sizeof])[0];
+
+    version (LittleEndian)
     {
-        import mir.ion.conv: serde;
-        import msgpack: unpack;
-        return data.unpack.value.serde!T;
+        import core.bitop : bswap, byteswap;
+        static if (T.sizeof >= 4) {
+            ret = bswap(ret);
+        } else static if (T.sizeof == 2) {
+            ret = byteswap(ret);
+        }
     }
 
-    version(mir_ion_test)
-    unittest
+    return cast(typeof(return))ret;
+}
+
+private static size_t readMap(S)(ref S serializer, scope const(ubyte)[] data, size_t length)
```

Comment:
For MsgPack we can pass data by reference:
```
private void readMap(S)(ref S serializer, ref scope const(ubyte)[] data, size_t length)
```
instead of returning length. 

In all of the API.

Reply:
Sounds good. I just implemented this - let me know what you think.